src/nlr/DeepPolySignElement.cpp

```diff
@@ -0,0 +1,231 @@
+/*********************                                                        */
+/*! \file DeepPolySignElement.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Haoze Andrew Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#include "DeepPolySignElement.h"
+#include "FloatUtils.h"
+
+namespace NLR {
+
+DeepPolySignElement::DeepPolySignElement( Layer *layer )
+{
+    _layer = layer;
+    _size = layer->getSize();
+    _layerIndex = layer->getLayerIndex();
+}
+
+DeepPolySignElement::~DeepPolySignElement()
+{
+    freeMemoryIfNeeded();
+}
+
+void DeepPolySignElement::execute( const Map<unsigned, DeepPolyElement *>
+                               &deepPolyElementsBefore )
+{
+    log( "Executing..." );
+    ASSERT( hasPredecessor() );
+    allocateMemory();
+
+    // Update the symbolic and concrete upper- and lower- bounds
+    // of each neuron
+    for ( unsigned i = 0; i < _size; ++i )
+    {
+        NeuronIndex sourceIndex = *( _layer->getActivationSources( i ).begin() );
+        DeepPolyElement *predecessor =
+            deepPolyElementsBefore[sourceIndex._layer];
+        double sourceLb = predecessor->getLowerBound
+            ( sourceIndex._neuron );
+        double sourceUb = predecessor->getUpperBound
+            ( sourceIndex._neuron );
+
+        if ( !FloatUtils::isNegative( sourceLb ) )
+        {
+            // Phase positive
+            // Symbolic bound: 1 <= x_f <= 1
+            // Concrete bound: 1 <= x_f <= 1
+            _symbolicUb[i] = 0;
+            _symbolicUpperBias[i] = 1;
+            _ub[i] = 1;
+
+            _symbolicLb[i] = 0;
+            _symbolicLowerBias[i] = 1;
+            _lb[i] = 1;
+        }
+        else if ( !FloatUtils::isPositive( sourceUb ) )
```

Comment:
Signs are not continuous. We're in the negative phase if the input is negative, and in the positive phase if it's non negative. Better change this to isNegative (instead of !isPositive)